### PR TITLE
Merge pull request #2 from queens-supercluster/main

### PR DESCRIPTION
**Pull request by erinCodes3276**
Merging endpoints

### CODE REVIEW

**Comment by erinCodes3276** [comment]
merge

File: prewl/endpoints/gpt2.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        import tensorflow as tf
+        tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.ERROR)
+        from transformers import TFGPT2LMHeadModel, GPT2Tokenizer
+        tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
+        model = TFGPT2LMHeadModel.from_pretrained("gpt2", pad_token_id=tokenizer.eos_token_id)
+
+        # encode context the generation is conditioned on
+        input_ids = tokenizer.encode(prompt, return_tensors='tf')
+
+        # generate text until the output length (which includes the context length) reaches 50
```
Comment (haz) [inline prewl/endpoints/gpt2.py:17]: Comment needs to be updated.

File: prewl/endpoints/gpt2.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        greedy_output = model.generate(input_ids, max_length=100)
```
Comment (haz) [inline prewl/endpoints/gpt2.py:18]: The `max_length` should come from the prewl settings.

Are there any other settings we might use here or in the next line?

File: run_file.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import prewl, json
```
Comment (haz) [inline run_file.py:1]: This file shouldn't be included in the repo...just for your personal testing. It's all in the README already.

If you really want short snippets to demonstrate different functionality, you can create an `examples/` directory.

**Review by haz — CHANGES_REQUESTED** [review]
Thanks! It still seems to be giving some wonky behaviour, though...

```bash
$ python3 run_file.py
All model checkpoint layers were used when initializing TFGPT2LMHeadModel.

All the layers of TFGPT2LMHeadModel were initialized from the model checkpoint at gpt2.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFGPT2LMHeadModel for predictions without further training.

 New input:  It was awesome!
Prediction:  Text: I really don't like this movie.
Sentiment: negative

Text: This flick was sick!
Sentiment: positive

Text: It was ok. I've seen better, though.
Sentiment: neutral

Text: It was awesome!
Sentiment: ive seen better.
Sentiment: ive seen better.
Sentiment: ive seen better.
Sentiment: ive seen better.
Sentiment: ive seen
```

In the `run_file.py` I commented out this block:

```python
def check_for_positive(resp):
    return "positive" in resp
model = prewl.train(prompts, check_for_positive)
```

This means that it's open-ended as to what gets returned. From the above, it's probably ` ive seen better.` that we want to see returned. There are two issues currently...

1. The full input prompt seems to be there, and that should be removed.
2. The LLM just carries on with the output (new lines, new prompts, etc).

I think both of these will require changes to the call function here.

Can you tell me what `greedy_output` is made up of?

File: run_file.py (hunk 3 above)
Comment (erinCodes3276) [reply run_file.py:1]: Just for testing purposes I should move this file then? Should the examples/ directory be in PREWL?

Comment (haz) [reply run_file.py:1]: Yes, this file should appear in the root directory. If you wanted to do the `examples/` directory, then you could place something like this in there. Small test files to show different functionality (one for gpt2, one for the callback function to check the output, etc).

File: libcuda.so.1 [hunk 4]
```diff
@@ -0,0 +1 @@
+libcuda.so.7.*
```
Comment (haz) [inline libcuda.so.1:1]: This file shouldn't be included.

File: PREWL [hunk 5]
```diff
@@ -0,0 +1,71 @@
+import prewl, json
```
Comment (haz) [inline PREWL:1]: What's this file?

File: prewl/endpoints/gpt2.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
+        
+        # Add the EOS token as PAD token to avoid warnings.
+        model = TFGPT2LMHeadModel.from_pretrained("gpt2", pad_token_id=tokenizer.eos_token_id)
+
+        # Encode context the text generation is conditioned to.
+        input_ids = tokenizer.encode(prompt, return_tensors='tf')
+
+        # Generates text until the output length using greedy search as the decoding method.
+        greedy_output = model.generate(input_ids, max_length=100)
```
Comment (haz) [inline prewl/endpoints/gpt2.py:22]: Have you tried pulling in the configuration yet?

**Review by haz — CHANGES_REQUESTED** [review]

File: libcuda.so.1 (hunk 4 above)
Comment (erinCodes3276) [reply libcuda.so.1:1]: I removed it. I think I accidentally added it while trying to figure out how to connect gpt2

File: PREWL (hunk 5 above)
Comment (erinCodes3276) [reply PREWL:1]: It seems to be a copy of the run_file file I made.

File: prewl/endpoints/gpt2.py (hunk 6 above)
Comment (erinCodes3276) [reply prewl/endpoints/gpt2.py:22]: I'm not sure what file to  add the max_length configuration to

Comment (haz) [reply prewl/endpoints/gpt2.py:22]: You should be able to load the configuration with `from prewl import CONFIG`, and then you should be able to access the max length with `CONFIG['max-length']` -- just make sure you include that `max-length` in the configuration (top of the example file).

File: PREWL (hunk 5 above)
Comment (haz) [reply PREWL:1]: Should be removed for this PR to go in.

**Review by haz — CHANGES_REQUESTED** [review]
Any update on the other stuff?

File: run_file.py (hunk 3 above)
Comment (erinCodes3276) [reply run_file.py]: I created an examples directory and put run_file in it. I will change its name to show that it is testing gpt2

File: PREWL (hunk 5 above)
Comment (erinCodes3276) [reply PREWL]: removed now

**Comment by haz** [comment]
Merging so I can polish things up a bit for a working installable lib.

**Pull request merged**